Software Versions: Add support to all modules

pontushojer opened the pull request:
This PR works towards solving https://github.com/ewels/MultiQC/issues/2042

Version parsing added for modules:

- adaptorremoval
- anglerfish
- bakta
- bbduk
- bcftools
- bismark
- cellranger
- clipandmerge
- cutadapt
- damageprofilier
- dedup
- diamond
- eigenstratdatabasetools
- fastq_screen
- fastqc
- flash
- flexbar
- gffcompare
- gopeaks
- hifiasm
- interop
- longranger
- macs2
- malt
- mirtop
- mtnucratio
- multivcfanalyzer
- pangolin
- purple
- pycoqc
- qc3C
- salmon
- sambamba

---

- [x] This comment contains a description of changes (with reason)
- [ ] `CHANGELOG.md` has been updated

vladsavelyev commented:
Wow, that's really amazing!! Thanks so much for doing this, @pontushojer

vladsavelyev commented on multiqc/modules/cutadapt/cutadapt.py:
```diff
@@ -100,6 +100,7 @@ def parse_cutadapt_logs(self, f):
         s_name = None
         end = "default"
         cutadapt_version = None
+        python_version = None
```
I think we should only report versions of the scientific software, so python version probably is not important

vladsavelyev reviewed (approved):
It looks great, approved!

Some late thoughts after looking at the table again, should have been on the original PR, but 😅 
* I'm still not convinced that reporting `htslib` versions for `bcftools` and `samtools` is a useful information, as almost always they going to be identical to the actual tools versions. I'd remove it, or at least add some logic to only show it if it differs 🤔  And similarly, no need to report the Python version for cutadapt.
* Also, another late thought I don't think we should lowercase module names in the table. It's nicer when the names match the sections names in the report (e.g. "Adapter Removal" instead of "adapter removal"). And no need to wrap module names in the code tags.

ewels commented:
@pontushojer - you're a machine! 🤯  Amazing work! 👏🏻 

@vladsavelyev - agree on both points. I realise that I pushed for the htslib thing 👀 but I also like clean output 😅

pontushojer commented:
Thanks guys! 

> * I'm still not convinced that reporting `htslib` versions for `bcftools` and `samtools` is a useful information, as almost always they going to be identical to the actual tools versions. I'd remove it, or at least add some logic to only show it if it differs 🤔  And similarly, no need to report the Python version for cutadapt.

From my perspective I would assume its useful information if the tool author(s) decides to include it in their log output. Thus I would say its probably better to include it than not. Probably samtools can be compiled with different versions of htslib thus it is valuable to report this.

> * Also, another late thought I don't think we should lowercase module names in the table. It's nicer when the names match the sections names in the report (e.g. "Adapter Removal" instead of "adapter removal"). And no need to wrap module names in the code tags.

This I agree would be a nice change. I will open a separate PR with this.

vladsavelyev commented on multiqc/modules/bcftools/stats.py:
```diff
@@ -41,6 +45,16 @@ def parse_bcftools_stats(self):
+
+                    htslib_version_match = re.search(HTSLIB_REGEX, line)
+                    if htslib_version_match is not None:
+                        self.add_software_version(htslib_version_match.group(1), f["s_name"], "htslib")
```
```suggestion
                        if htslib_version_match.group(1) != version_match.group(1):
                            self.add_software_version(htslib_version_match.group(1), f["s_name"], "htslib")
```

vladsavelyev replied:
I'd suggest to only report it if it's different. What do you think?

ewels commented:
> Probably samtools can be compiled with different versions of htslib thus it is valuable to report this.

How about we add some code in the samtools module which checks if samtools == htslib. If they're different, report both as done currently. If they're the same, just report samtools (can always do a debug log message if we're feeling paranoid).

Then we get clean output in the 99.99% of cases, but in the edge cases we still show the important information.

pontushojer commented:
@ewels, @vladsavelyev I changed both samtools and bcftools to only report htslib version if its different. 

What do you think about the `cutadapt` case? This can definitely be run using different version of python so I think its best to include. 

Also I now added `pagolin` ([c595cad](https://github.com/ewels/MultiQC/pull/2051/commits/c595cade13ef903dd4b6a47a91ee99fd241f5128)) which reports a whole array of software version for tools it depends on. Not sure if all are relevant but maybe that is for someone more familiar with the tool to decide whats should be included. For now I added most of them.

ewels commented:
I think Python version is too deep. It's not directly related / _part of_ the tool, but rather it's basically a dependency. Let's keep it to just the cutadapt version.

Pangolin - haven't checked what the output looks like, but here I'm more tempted report everything. This tool is made of several subcomponents and stuff like the database version etc is very important. But I can't remember what is already reported and the details of what you've put here, so I'll check properly later. Maybe we can ask on the @nf-core viralrecon slack channel if we're not sure.

ewels commented:
Yeah, Pangolin is kind of scrappy but I think that it's probably fairly unique. That information is important, so let's keep it like that for now. Can always revise according to community feedback.

ewels reviewed (approved):
Amazing work @pontushojer ! 👏🏻 

Going to merge now as we are building towards a release. But keep 'em coming! 😆

pontushojer commented:
> Yeah, Pangolin is kind of scrappy but I think that it's probably fairly unique. That information is important, so let's keep it like that for now. Can always revise according to community feedback.

Sound good to me! 

I was also looking into the testdata we have for Pangolin in `MultiQC_TestData` . It seems like that columns are left-shifted for some sample which makes version end up in the wrong columns: see https://github.com/ewels/MultiQC_TestData/blob/master/data/modules/pangolin/v4.0/pangolin_4-0.csv

Posted an issue here: https://github.com/cov-lineages/pangolin/issues/531. Let's see if it is just a misstake or might occur in the actual tool output.
